eth_account/signers/ledger.py

```diff
@@ -0,0 +1,362 @@
+
+    # Split command into at max PACKET_FREE sized chunks
+    commands = [command[i:i + PACKET_FREE] for i in range(0, len(command), PACKET_FREE)]
+
+    # Create a packet for each command chunk
+    for packet_id in range(len(commands)):
+        header = struct.pack('>HBH', CHANNEL_ID, TAG_APDU, packet_id)
+        packet = header + commands[packet_id]
+
+        # Add padding to the packet to make it exactly PACKET_SIZE long
+        packet += bytes([0x0] * (PACKET_SIZE - len(packet)))
+
+        packets.append(packet)
+
+    return packets
+
+
+def unwrap_apdu(packet):
+    '''
+    Given a packet from the device, extract and return relevant info
+    '''
+    if not packet:
+        return (None, None, None, None, None)
+
+    (channel, tag, packet_id, reply_size) = struct.unpack('>HBHH', packet[:7])
+
+    if packet_id == 0:
+        # reply_size is only valid in first reply
+        return (channel, tag, packet_id, reply_size, packet[7:])
+    else:
+        return (channel, tag, packet_id, None, packet[5:])
+
+
+class LedgerUsbException(Exception):
+    pass
+
+
+class LedgerUsbDevice:
+    '''
+    References:
+    - https://github.com/LedgerHQ/blue-loader-python/blob/master/ledgerblue/comm.py#L56
+    - https://github.com/ethereum/go-ethereum/blob/master/accounts/usbwallet/ledger.go
+    '''
+
+    logger = logging.getLogger('eth_account.signers.ledger.LedgerUsbDevices')
+
+    def __init__(self):
+        hidDevicePath = None
+        for hidDevice in hid.enumerate(0, 0):
+            if hidDevice['vendor_id'] == 0x2c97:
+                if ('interface_number' in hidDevice and hidDevice['interface_number'] == 0) \
+                   or ('usage_page' in hidDevice and hidDevice['usage_page'] == 0xffa0):
+                    hidDevicePath = hidDevice['path']
+        if hidDevicePath is not None:
+            dev = hid.device()
+            dev.open_path(hidDevicePath)
+            dev.set_nonblocking(True)
+        else:
+            raise LedgerUsbException('No Ledger usb device found')
+        self.device = dev
+
+    def exchange(self, apdu, timeout=20):
+        self.logger.debug(f'Sending apdu to Ledger device: {to_hex(apdu)}')
+
+        # Construct the wrapped packets
+        packets = wrap_apdu(apdu)
+
+        # Send to device
+        for packet in packets:
+            self.device.write(packet)
+
+        # Receive reply, size of reply is contained in first packet
+        reply = []
+        reply_min_size = 2
+        reply_start = time.time()
+        while True:
+            packet = bytes(self.device.read(64))
+            (channel, tag, index, size, data) = unwrap_apdu(packet)
+
+            # Wait for a valid channel in replied packet
+            if not channel:
+                if reply_start + timeout < time.time():
+                    raise LedgerUsbException(f'Timeout waiting for a device' +
+                                             f'response (timeout={timeout}s)')
```

Comment:
TIL about f-strings!

Reply:
Yeah, they're pretty cool. I can't wait to use them more broadly! Unfortunately f-strings weren't added until 3.6, but Web3.py (and therefore eth-account) has to support py3.5. So this will have to be switched out for `%` or `.format` style instead. :(

Reply:
That explain why circleCI lint was failing ! Too bad we are not on 3.6 :)
I will update to remove all f-strings.